### PR TITLE
Ios dedup check added

### PR DESCRIPTION
**Pull request by ayush03agarwal**

### CODE REVIEW

**Comment by kingster** [comment]
make changes as discussed

File: receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala [hunk 1]
```diff
@@ -43,22 +49,33 @@ class RegistrationRoute(implicit am: ActorMaterializer) extends BaseJsonHandler
...
+                              val newDeviceDetails = d.copy(appName = appName, osName = platform.toString, deviceId = deviceId, active = true)
+                              newDeviceDetails.validate()
+                              if (!isTestRequest) {
 
-                              result match {
-                                case Right(x) =>
-                                  complete(GenericResponse(StatusCodes.Created.intValue, null, Response(s"DeviceDetails created for ${newDeviceDetails.deviceId}", newDeviceDetails)))
-                                case Left(x) =>
-                                  complete(GenericResponse(StatusCodes.OK.intValue, null, Response(s"DeviceDetails updated for ${newDeviceDetails.deviceId}", newDeviceDetails)))
+                                val getDeviceDetailsWithId = Future(DeviceDetailsService.get(appName, deviceId).get)
```
Comment (kingster) [inline receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala]: probably betterNamed as existingDeviceDetails

File: receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala [hunk 2]
```diff
@@ -43,22 +49,33 @@ class RegistrationRoute(implicit am: ActorMaterializer) extends BaseJsonHandler
...
+                                val getDeviceDetailsWithToken = Future(DeviceDetailsService.getByTokenId(newDeviceDetails.appName, newDeviceDetails.token).get)
```
Comment (kingster) [inline receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala]: Variables shouldn't start with get****

File: receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala [hunk 3]
```diff
@@ -43,22 +49,33 @@ class RegistrationRoute(implicit am: ActorMaterializer) extends BaseJsonHandler
...
+
+                                getDeviceDetailsWithToken.flatMap[ToResponseMarshallable] {
+                                  case Some(device) if device.deviceId != newDeviceDetails.deviceId =>
+                                    meter("invalid.deviceDetails").mark()
+                                    ConnektLogger(LogFile.SERVICE).error(s"DeviceDetails add/update failed for ${newDeviceDetails.deviceId} as token ${newDeviceDetails.token} is already assigned to deviceId ${device.deviceId}")
+                                    Future(GenericResponse(StatusCodes.BadRequest.intValue, null, Response(s"DeviceDetails add/update failed for ${newDeviceDetails.deviceId} as token ${newDeviceDetails.token} is already assigned to deviceId ${device.deviceId}", null)).respond)
```
Comment (kingster) [inline receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala]: Return FastFuture here. No need to create a future and pass context

File: receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala [hunk 4]
```diff
@@ -43,22 +49,33 @@ class RegistrationRoute(implicit am: ActorMaterializer) extends BaseJsonHandler
...
+                                  case _ =>
+                                    getDeviceDetailsWithId.flatMap[ToResponseMarshallable] {
+                                      case Some(deviceDetail) =>
+                                        DeviceDetailsService.update(deviceId, newDeviceDetails)
+                                        Future(GenericResponse(StatusCodes.OK.intValue, null, Response(s"DeviceDetails updated for ${newDeviceDetails.deviceId}", newDeviceDetails)).respond)
```
Comment (kingster) [inline receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala]: Same, use fastfuture

File: receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala [hunk 5]
```diff
@@ -43,22 +49,33 @@ class RegistrationRoute(implicit am: ActorMaterializer) extends BaseJsonHandler
...
+                                      case None =>
+                                        DeviceDetailsService.add(newDeviceDetails)
+                                        Future(GenericResponse(StatusCodes.Created.intValue, null, Response(s"DeviceDetails created for ${newDeviceDetails.deviceId}", newDeviceDetails)).respond)
```
Comment (kingster) [inline receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala]: Fasfuture

File: receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala [hunk 6]
```diff
@@ -43,22 +49,33 @@ class RegistrationRoute(implicit am: ActorMaterializer) extends BaseJsonHandler
...
+                                    }
+                                }
+                              }
+                              else {
+                                GenericResponse(StatusCodes.Created.intValue, null, Response(s"DeviceDetails skipped for ${newDeviceDetails.deviceId}", newDeviceDetails)).respond
```
Comment (kingster) [inline receptors/src/main/scala/com/flipkart/connekt/receptors/routes/master/RegistrationRoute.scala]: Why mix future and no-future?

**Comment by ayush03agarwal** [comment]
changes done

**Pull request merged**